kartothek/io_components/read.py

```diff
@@ -47,14 +48,28 @@ def dispatch_metapartitions_from_factory(
         if isinstance(ix, ExplicitSecondaryIndex)
     }
 
-    if concat_partitions_on_primary_index:
+    if dispatch_by and concat_partitions_on_primary_index:
+        raise ValueError(
+            "Both `dispatch_by` and `concat_partitions_on_primary_index` are provided, "
+            "`concat_partitions_on_primary_index` is deprecated and will be removed in the next major release. "
+            "Please only provide the `dispatch_by` argument. "
+        )
+
+    if dispatch_by or concat_partitions_on_primary_index:
         if dataset_factory.explicit_partitions:
             dataset_factory = dataset_factory.load_partition_indices()
 
+        if dispatch_by:
+            if not set(dispatch_by).issubset(dataset_factory.partition_keys):
+                raise ValueError("Nope.")
```

Comment:
With the logical predicates you can relax this test to: all of the columns need to be an index column (not just restricted to partition cols)